src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -181,234 +209,248 @@ private void CreateStreamAndStartWatcher()
+                    // First, we should check if this event should kick off a re-scan since we can't really rely on anything after this point if that is true
+                    if (ShouldRescanOccur(eventFlags[i]))
                     {
-                        // Find the rename that is paired to this rename, which should be the next rename in the list
-                        long pairedId = FindRenameChangePairedChange(i, eventPaths, eventFlags, eventIds);
-                        if (pairedId == long.MinValue)
+                        watcher.OnError(new ErrorEventArgs(new IOException(SR.FSW_BufferOverflow, (int)eventFlags[i])));
+                        break;
+                    }
+                    else if ((handledRenameEvents != null) && (handledRenameEvents.Contains(eventIds[i])))
+                    {
+                        // If this event is the second in a rename pair then skip it
+                        continue;
+                    }
+                    else if ((CheckIfPathIsNested(path)) && ((_filterFlags & eventFlags[i]) != 0))
+                    {
+                        // The base FileSystemWatcher does a match check against the relative path before combining with 
+                        // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
+                        string relativePath = null;
+                        if (eventPaths[i].Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase) == false)
                         {
-                            // Getting here means we have a rename without a pair, meaning it should be a create for the 
-                            // move from unwatched folder to watcher folder scenario or a move from the watcher folder out.
-                            // Check if the item exists on disk to check which it is
-                            if (DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)))
+                            // Check if the event path, with the root directory removed, begins with a / and
+                            // if so, remove it; otherwise, just remove the root path (which contains a trailing /)
+                            if (eventPaths[i][_fullDirectory.Length] == '/')
                             {
-                                NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
+                                relativePath = eventPaths[i].Remove(0, _fullDirectory.Length + 1);
                             }
                             else
                             {
-                                NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
+                                relativePath = eventPaths[i].Remove(0, _fullDirectory.Length);
```

Comment:
I don't mind the extra verbosity here since it makes it super clear what is going on...if you feel strongly I can change it, but I'd rather the expanded version here in order to keep readability